working/0107 - implicit-constructors/feature-specification.md

```diff
@@ -0,0 +1,603 @@
+
+<staticExtensionMethodSignature> ::= // New rule.
+    <functionSignature> |
+    <getterSignature> |
+    <setterSignature>;
+
+<staticExtensionConstructor> ::= // New rule.
+    <factoryConstructorSignature> <functionBody> |
+    <redirectingFactoryConstructorSignature> ';';
+
+<staticExtensionVariableDeclaration> ::= // New rule.
+    ('final' | 'const') <type>? <staticFinalDeclarationList> |
+    'late' 'final' <type>? <initializedIdentifierList> |
+    'late'? <varOrType> <initializedIdentifierList>;
+```
+
+The identifier `implicit` is now mentioned in the grammar, but it is not a
+built-in identifier nor a reserved word. *The parser does not need that.*
+
+In a static extension of the form `static extension E on C {...}` where `C`
+is an identifier or an identifier with an import prefix, we say that the
+on-class of the static extension is `C`. If `C` resolves to a non-generic
+class then we say that the _constructor return type_ of the static
+extension is `C`.
+
+*If `C` resolves to a generic class then the static extension does not have
+a constructor return type.*
+
+In a static extension of the form `static extension E on C<T1 .. Tk> {...}`
+where `C` is an identifier or prefixed identifier, we say that the on-class
+of `E` is `C`, and the _constructor return type_ of `E` is `C<T1 .. Tk>`.
+
+In both cases, `E` is an identifer `id` which is optionally followed by a
+term derived from `<typeParameters>`. We say that the identifier `id` is
+the _name_ of the static extension. *Note that `T1 .. Tk` above may contain
+occurrences of those type parameters.*
+
+### Static Analysis
+
+At first, we establish some sanity requirements for a static extension
+declaration by specifying several errors.
+
+A compile-time error occurs if the on-class of a static extension does not
+resolve to an enum declaration or a declaration of a class, a mixin, a mixin
+class, or an inline class.
+
+A compile-time error occurs if a static extension has an on-clause of the form
+`on C` where `C` denotes a generic class and no type arguments are passed to `C`
+*(i.e., it is a raw type)*, and the static extension contains one or more
+constructor declarations.
+
+*In other words, if the static extension ignores the type parameters of the
+on-class then it can only contain `static` members. Note that if the
+on-class is non-generic then it is not a raw type, and the static extension
+can contain constructors.*
+
+A compile-time error occurs if a static extension has an on-clause of the
+form `on C<T1 .. Tk>`, and the actual type parameters passed to `C` do not
+satisfy the bounds declared by `C`.  The static extension may declare one
+or more type parameters `X1 extends B1 .. Xs extends Bs`, and these type
+variables may occur in the types `T1 .. Tk`. During the bounds check, the
+bounds `B1 .. Bs` are assumed for the type variables `X1 .. Xs`.
+
+A compile-time error occurs if a static extension _D_ declares a
```

Comment:
This makes it a breaking change to add a constructor to a class.  Probably worth at least calling this out for consideration?  It might be better to do what we do with extension methods and just say that the class version takes precedence (or the other way around, I suppose).

Reply:
It is no more a breaking change.